### PR TITLE
Optimize (hot) Symbol#name, and some other things

### PR DESCRIPTION
**Pull request by hrhino**
I see a (somewhat shaky) [improvement of ~1%](https://gist.github.com/64d1dd3554dfad98b2e8212667386c87) on this; although I haven't benchmarked individual commits, I suspect that the `Symbol#name` change is responsible for the lion's share of it.

cc @mkeskells @rorygraves

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 1]
```diff
@@ -242,16 +243,15 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
     // Rename as little as possible.  Enforce invariants on all renames.
     type TypeOfClonedSymbol >: Null <: Symbol { type NameType = Symbol.this.NameType }
 
-    // Abstract here so TypeSymbol and TermSymbol can have a private[this] field
-    // with the proper specific type.
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala:246]: 👍 for this change. I've tried variations of this myself before.

I think the justification in this removed comment was optimizing for the wrong thing.

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 2]
```diff
@@ -242,16 +243,15 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
-    def rawname: NameType
-    def name: NameType
+    final def rawname: NameType = _rawname.asInstanceOf[NameType]
+    final def name: NameType = if (needsFlatClasses) flattenedName else _rawname.asInstanceOf[NameType]
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala:247]: Let's try to add a faster path for `MethodSymbol` (and for `TermSymbol` with the `METHOD` flag, which I think is what [we get from unpickling](https://github.com/scala/scala/blob/2d9e6acce8c4246e6cba5a22ff9d965ec3bd117c/src/reflect/scala/reflect/internal/pickling/UnPickler.scala#L363)),  that avoids the virtual call to `flattenedName`. It's probably enough to add another flag test to `needsFlatClasses`.

Comment (hrhino) [reply src/reflect/scala/reflect/internal/Symbols.scala:247]: That makes sense. I looked at `needsFlatClasses` and I don't think there's anything that calls it on a method expecting the answer to be `true`.

**Comment by SethTisue** [comment]
@retronym merge?

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 3]
```diff
@@ -840,7 +840,7 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
     final def isDelambdafyFunction = isSynthetic && (name containsName tpnme.DELAMBDAFY_LAMBDA_CLASS_NAME)
     final def isDelambdafyTarget  = isArtifact && isMethod && hasAttachment[DelambdafyTarget.type]
     final def isDefinedInPackage  = effectiveOwner.isPackageClass
-    final def needsFlatClasses    = phase.flatClasses && (rawowner ne NoSymbol) && !rawowner.isPackageClass
+    final def needsFlatClasses    = !isMethod && phase.flatClasses && (rawowner ne NoSymbol) && !rawowner.isPackageClass
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala]: `isMethod` is actually a virtual call as things are currently factored. We could push this check to the end of this line (typically `phase.flatClasses` is mostly false).

As a followup task, we should review the way `isMethod` etc etc are factored with the help of micro-benchmarks and JITWatch analysis.

Comment (retronym) [reply src/reflect/scala/reflect/internal/Symbols.scala]: I've pushed the condition reordering as an extra commit to this PR.

**Review by retronym — APPROVED** [review]

**Pull request merged**